interp/constrintern.ml

```diff
@@ -2883,13 +2884,20 @@ let interp_context_evars_gen ?(program_mode=false) ?(unconstrained_sorts = false
           match b' with
           | None ->
             let int_env = if autoimp_enable then Name.fold_left (push_auto_implicit env sigma t) int_env na else int_env in
-            let d = make_decl (LocalAssum (make_annot na r,t)) in
+            let sigma, d =
+              match with_instances with
+              | Some (src, typeclass_candidate) ->
+                let sigma, ev = Evarutil.new_evar env sigma ~src:((Loc.tag ?loc @@ Evar_kinds.GoalEvar)) ~typeclass_candidate t in
+                sigma, make_decl ?loc (LocalDef (make_annot na r,ev,t))
+              | None ->
+                sigma,make_decl ?loc (LocalAssum (make_annot na r,t))
+            in
```

Comment:
We could also warn about ignored implicit information around that code

Reply:
I agree that it is ad hoc.

A related question is whether we want to accept binders of the form `(x:=t)` in `Derive`, that is, whether we want to support things like:
```
Derive x (y:=2) z in (x+y=z) as spec.
```
I started to support it (and the PR currently supports it) but I'm now less convinced. That might be clearer to only allow binders of the form `x` or `x : T`. I don't know your opinion.

Reply:
Let's start with the simpler syntax, if we want more later we can change it.

Reply:
I made the following compromise:
- I kept relying on the syntax of binders (because it is already done, and it is not clear yet that there will not be further discussions; moreover, it would be easy to remove local defs in `Derive.check_allowed_binders` if one wants)
- I let `derive.ml` instantiate the assumptions by evars, renouncing to introduce `interp_named_context_evars_as_arguments`